Update serverless custom metrics doc (#10977)
DataDog/documentation: content/en/serverless/custom_metrics/_index.md

kind: documentation
---

## Overview

There are a few different ways to submit custom metrics to Datadog from a Lambda function. 

- **Creating custom metrics from logs or traces**: If your Lambda functions are already sending trace or log data to Datadog, and the data you want to query is captured in an existing log or trace, you can [generate custom metrics from logs and traces](#creating-custom-metrics-from-logs-or-traces) without re-deploying or making any changes to your application code.

> - **Submitting custom metrics using the Datadog Lambda Extension**: If you want to submit custom metrics directly from your Lambda function, Datadog recommends using the [Datadog Lambda Extension](#with-the-datadog-lambda-extension). Check whether the Datadog Lambda Extension is supported for the runtime of your Lambda function [here][1].  

Reviewer: ```suggestion
- **Submitting custom metrics using the Datadog Lambda Extension**: If you want to submit custom metrics directly from your Lambda function, Datadog recommends using the [Datadog Lambda Extension](#with-the-datadog-lambda-extension). [Check whether the Datadog Lambda Extension is supported][1] in your Lambda function runtime.  
```

Reviewer: cc @tianchu I'm told to avoid using "here" for links for a few reasons:
- it's wordy, and you could just tie the link to the action
- it's not as good for screen readers

- **Submitting custom metrics using the Datadog Forwarder Lambda**: If you want to submit custom metrics from a runtime that is not yet supported by the Datadog Lambda Extension, you can use the [Datadog Forwarder Lambda](#with-the-datadog-forwarder).
- **(Deprecated) Submitting custom metrics from CloudWatch logs**: The method to submit custom metrics by printing a log formatted as `MONITORING|<UNIX_EPOCH_TIMESTAMP>|<METRIC_VALUE>|<METRIC_TYPE>|<METRIC_NAME>|#<TAG_LIST>` has been [deprecated](#deprecated-cloudwatch-logs), and you should migrate to one of the solutions above.
- **(Not recommended) Using a third-party library**: Most third-party libraries do not submit metrics as distributions and can lead to under-counted results.

### Understanding distribution metrics

> Custom metrics submitted from Lambda functions are aggregated as [distributions][2], because they are designed to instrument applications, independent of the underlying hosts. You can query the metrics using aggregations: `avg`, `sum`, `max`, `min`, `count`. You can also enable percentile aggregations (p50, p75, p90, p95, p99), and [manage tags][3] for aggregation on the Metric Summary page.

Reviewer: > because they are designed to instrument applications, independent of the underlying hosts.

What does this mean? I'm not sure most customers will understand this?

> Custom metrics submitted from Lambda functions are aggregated as [distributions][2], because they are designed to instrument applications, independent of the underlying hosts. You can query the metrics using aggregations: `avg`, `sum`, `max`, `min`, `count`. You can also enable percentile aggregations (p50, p75, p90, p95, p99), and [manage tags][3] for aggregation on the metric summary page.

Reviewer: > because they are designed to instrument applications, independent of the underlying hosts.

What do you mean by this?

Reviewer: ```suggestion
Custom metrics submitted from Lambda functions are aggregated as [distributions][2], because they are designed to instrument applications, independent of the underlying hosts. You can query the metrics using aggregations: `avg`, `sum`, `max`, `min`, `count`. You can also enable percentile aggregations (p50, p75, p90, p95, p99), and [manage tags][3] for aggregation on the Metric Summary page.
```

## Creating custom metrics from logs or traces

With log-based metrics, you can record a count of logs that match a query or summarize a numeric value contained in a log, such as request duration. Log-based metrics are a cost-efficient way to summarize log data from the entire ingest stream. Learn more about creating log-based metrics [here][4]. 

You can also generate metrics from 100% of ingested spans, regardless of whether they are indexed by a retention filter. Learn more about creating span-based metrics [here][5].
## With the Datadog Lambda Extension

{{< img src="serverless/serverless_custom_metrics.png" alt="Collecting Custom Metrics from AWS Lambda" >}}

Datadog recommends using the [Datadog Lambda Extension][1] to submit custom metrics from supported Lambda runtimes.

1. Follow the general [serverless installation instructions][6] to configure your Lambda function and install the Datadog Lambda Library and Extension.

> 1. If you are not interested in collecting traces from your Lambda function, set the environment variable `DD_TRACE_ENABLED` to `false`.

Reviewer: Should this be 2. instead of 1?

Author: The best practice is to always use `1` and let markdown automatically render the right number.

Reviewer 2: this is a markdown formatted ordered list and our site supports it (note that it's rendering correctly in the preview), so we can leave it this way

1. If you are not interested in collecting logs from your Lambda function, set the environment variable `DD_SERVERLESS_LOGS_ENABLED` to `false`.
1. Import and use the helper function from the Datadog Lambda Library, such as `lambda_metric` or `sendDistributionMetric`, to submit your custom metrics following the [sample code](#custom-metrics-sample-code).

> If your Lambda function is associated with a VPC, ensure that your function can reach Datadog API endpoints either through the public internet, [PrivateLink][7] or a [proxy][8].

Reviewer: ```suggestion
If your Lambda function is running in a VPC, ensure that your function can reach Datadog API endpoints either through the public internet, [PrivateLink][7] or a [proxy][8].
```

Reviewer: I've never heard "associated with" a VPC, so just double checking that's what you meant

Author: Haha, ask Stephen :)

## With the Datadog Forwarder

Datadog recommends using the [Datadog Forwarder Lambda][9] to submit custom metrics from Lambda runtimes that are not yet supported by the Datadog Lambda Extension.

1. Follow the general [serverless installation instructions][6] to configure your Lambda function, install the Datadog Lambda Library and the Datadog Forwarder Lambda function, and subscribe the Forwarder to your function's log group.

> 1. If you are not interested in collecting traces from your Lambda function, set the environment variable `DD_TRACE_ENABLED` to `false` on your own Lambda function.

Reviewer: Should this be 2. instead of 1?

kind: documentation
---

## Overview

There are a few different ways to submit custom metrics to Datadog from a Lambda function. 

- **Creating custom metrics from logs or traces**: If your Lambda functions are already sending trace or log data to Datadog, and the data you want to query is captured in an existing log or trace, you can [generate custom metrics from logs and traces](#creating-custom-metrics-from-logs-or-traces) without re-deploying or making any changes to your application code.

> - **Submitting custom metrics synchronously**: If you want to generate custom metrics from a non-performance-critical Lambda function written in Python, Node.js or Golang and you are not interested in collecting traces, you can use Datadog Lambda Library to [submit custom metrics synchronously](#enabling-synchronous-custom-metrics).

Reviewer 3: Should we be encouraging this option? Isn't the extension strictly better for this use case? I think it would be nice to eventually remove this feature.

Reviewer 3: At the very least, I think this should be listed below the asynchronous option, as that would be preferable for most customers.

Reviewer: We can put it a bit lower on the list with a (Not recommended) in front

Author: That's good feedback, actually I think we can remove this solution from the doc, the gap leaving behind seems minimal.

> - **Submitting custom metrics asynchronously**: If you want to generate custom metrics from a performance-critical Lambda function, Datadog encourage you to use Datadog Lambda Library to [submit custom metrics asynchronously](#enabling-asynchronous-custom-metrics).

Reviewer 2: ```suggestion
- **Submitting custom metrics asynchronously**: If you want to generate custom metrics from a performance-critical Lambda function, Datadog recommends using the Datadog Lambda Library to [submit custom metrics asynchronously](#enabling-asynchronous-custom-metrics).
```

> - **Submitting custom metrics asynchronously**: If you want to generate custom metrics from a performance-critical Lambda function, Datadog encourage you to use Datadog Lambda Library to [submit custom metrics asynchronously](#enabling-asynchronous-custom-metrics).

Reviewer 3: ```suggestion
- **Submitting custom metrics asynchronously**: If you want to generate custom metrics from a performance-critical Lambda function, Datadog encourages you to use the Datadog Lambda Library to [submit custom metrics asynchronously](#enabling-asynchronous-custom-metrics).
```

- **(Deprecated) Submitting custom metrics from CloudWatch logs**: The method to submit custom metrics by printing a log formatted as `MONITORING|<UNIX_EPOCH_TIMESTAMP>|<METRIC_VALUE>|<METRIC_TYPE>|<METRIC_NAME>|#<TAG_LIST>` has been [deprecated](#deprecated-cloudwatch-logs), and you should migrate to one of the solutions above.

> - **(Not recommended) Using a third-party library**: Most of [third-party libraries](#third-party-libraries) do not submit metrics as distributions and can lead to under-counted results.

Reviewer 2: ```suggestion
- **(Not recommended) Using a third-party library**: Most [third-party libraries](#third-party-libraries) do not submit metrics as distributions and can lead to under-counted results.
```

### Understanding distribution metrics

> Custom metrics submitted from Lambda functions are aggregated as [distributions][1], as they are designed to instrument applications, independent of the underlying hosts. You can query the metrics using aggregation `avg`, `sum`, `max`, `min`, `count`. You can also enable percentile aggregations (p50, p75, p90, p95, p99) and [manage tags][2] for aggregation on the metric summary page.

Reviewer 2: ```suggestion
Custom metrics submitted from Lambda functions are aggregated as [distributions][1], because they are designed to instrument applications, independent of the underlying hosts. You can query the metrics using aggregations: `avg`, `sum`, `max`, `min`, `count`. You can also enable percentile aggregations (p50, p75, p90, p95, p99), and [manage tags][2] for aggregation on the metric summary page.
```

> Certain submission methods allow you submit metrics with timestamps from the past, up-to 20 mins old.

Reviewer 3: When I read this, I immediately ask "which methods?". I think the answer is anything except the Extension. Since support for this is inconsistent, do we want to encourage it at all?

Author: Good feedback, let me call it out more explicitly.

## Creating custom metrics from logs or traces

With log-based metrics, you can record a count of logs that match a query or summarize a numeric value contained in a log, such as request duration. Log-based metrics are a cost-efficient way to summarize log data from the entire ingest stream. Learn more about creating log-based metrics [here][3]. 

You can also generate metrics from 100% of ingested spans, regardless of whether they are indexed by a retention filter. Learn more about creating span-based metrics [here][4].

## Enabling synchronous custom metrics

> If you want to generate custom metrics from a Lambda function that is written in Python, Node.js or Golang and is performance-critical, you can use Datadog Lambda Library to submit custom metrics synchronously to Datadog API at the end your Lambda invocation.

Reviewer 2: ```suggestion
If you want to generate custom metrics from a Lambda function that is written in Python, Node.js or Golang and is performance-critical, you can use the Datadog Lambda Library to submit custom metrics synchronously to Datadog API at the end of your Lambda invocation.
```

> If you want to generate custom metrics from a Lambda function that is written in Python, Node.js or Golang and is performance-critical, you can use Datadog Lambda Library to submit custom metrics synchronously to Datadog API at the end your Lambda invocation.

Reviewer 3: ```suggestion
If you want to generate custom metrics from a Lambda function that is written in Python, Node.js or Golang and is not performance-critical, you can use Datadog Lambda Library to submit custom metrics synchronously to Datadog API at the end your Lambda invocation.
```

Reviewer 3: I know that you meant to say "not performance-critical" here. However, even with that change, I think that makes it sound like if your Lambda is not performance-critical you **SHOULD** submit metrics synchronously, when in fact I think it would be better to use the Extension in all cases? Why would you ever want to submit metrics synchronously?

1. Follow the general [serverless installation instructions][5] to install the Datadog Lambda Library and configure your Lambda function. Note, you do _NOT_ need to add the Datadog Lambda Extension or Forwarder.
1. Ensure the environment variable `DD_FLUSH_TO_LOG` is set to `false`, which instructs the Datadog Lambda Library to send metrics synchronously to the Datadog API.
1. Ensure the environment variable `DD_TRACE_ENABLED` is set to `false`, which instructs the Datadog Lambda Library to NOT generate traces for the Lambda function.
1. Ensure the environment variable `DD_API_KEY` (or an equivalent one when encrypted) is set with a valid Datadog API key.
1. Import and use the helper function from the Datadog Lambda Library, such as `lambda_metric` or `sendDistributionMetric`, to submit your custom metrics following the [sample code](#custom-metrics-sample-code).
### Running in a VPC

> The Datadog Lambda Library requires [access to the public internet][6] to submit custom metrics **synchronously**. If your Lambda function is associated with a VPC, ensure that your function can reach the public internet.

Reviewer 3: This is also required to submit metrics asynchronously if you are using the Extension. I think the only case where it is not required is when using the Forwarder.

Author: Good point.

## Enabling asynchronous custom metrics

> If you want to generate custom metrics from a performance-critical Lambda function, Datadog encourage you to use Datadog Lambda Library to submit custom metrics asynchronously. The Datadog Lambda Library can either

Reviewer 2: ```suggestion
If you want to generate custom metrics from a performance-critical Lambda function, Datadog recommends to use the Datadog Lambda Library to submit custom metrics asynchronously. The Datadog Lambda Library can either:
```

> If you want to generate custom metrics from a performance-critical Lambda function, Datadog encourage you to use Datadog Lambda Library to submit custom metrics asynchronously. The Datadog Lambda Library can either

Reviewer 3: ```suggestion
If you want to generate custom metrics from a performance-critical Lambda function, Datadog encourages you to use Datadog Lambda Library to submit custom metrics asynchronously. The Datadog Lambda Library can either
```

Reviewer 3: It looks like this sentence is unfinished

Reviewer 3: OK, I think what you should change this as follows:

The Datadog Lambda Library can either**:** <-- add colon
- Publishes -> publish...
- Emits -> emit...

> - Publishes your custom metrics to the [Datadog Lambda Extension][7], which will submit your custom metrics directly to Datadog without impacting your function latency. This solution is easier to set up, but it currently only work for the popular runtimes, and does not support submitting metrics with timestamps from the past.

Reviewer 2: ```suggestion
- Publish your custom metrics to the [Datadog Lambda Extension][7], which submits your custom metrics directly to Datadog without impacting your function latency. This solution is easier to set up, but it currently only works for the popular runtimes, and does not support submitting metrics with timestamps from the past.
```
we should be more specific about what you mean by the popular runtimes

Author: Tried to be vague, because I know that list will quickly change in the future, and almost certain that we may forget updating the list here.

> - Publishes your custom metrics to the [Datadog Lambda Extension][7], which will submit your custom metrics directly to Datadog without impacting your function latency. This solution is easier to set up, but it currently only work for the popular runtimes, and does not support submitting metrics with timestamps from the past.

Reviewer 3: ```suggestion
- Publishes your custom metrics to the [Datadog Lambda Extension][7], which will submit your custom metrics directly to Datadog without impacting your function's latency. This solution is easier to set up, but it currently only works for the popular runtimes, and does not support submitting metrics with timestamps from the past.
```

Reviewer 3: For "the popular runtimes" -- which runtimes?

> - Emits your custom metrics as specially-formatted log which the [Datadog Forwarder][8] parses and submits to Datadog. This solution works for any runtime and metrics with timestamps from the past.

Reviewer 2: ```suggestion
- Emit your custom metrics as a specially-formatted log which the [Datadog Forwarder][8] parses and submits to Datadog. This solution works for any runtime and metrics with timestamps from the past.
```

> - Emits your custom metrics as specially-formatted log which the [Datadog Forwarder][8] parses and submits to Datadog. This solution works for any runtime and metrics with timestamps from the past.

Reviewer 3: ```suggestion
- Emits your custom metrics as specially-formatted logs which the [Datadog Forwarder][8] parses and submits to Datadog. This solution works for any runtime and supports metrics with timestamps from the past.
```

### With the Datadog Forwarder

You can emit metrics to CloudWatch logs using the Datadog Lambda Library.

1. Follow the general [serverless installation instructions][5] (certain runtimes default to using the Extension, and see [here][9] for the Forwarder-based installation guides) to configure your Lambda function, install the Datadog Lambda Library and the Datadog Forwarder Lambda function, and subscribe the Forwarder to your function's log group. For Lambda runtimes not listed in the installation guide, refer to the next section for solutions.
1. If you are not interested in collecting traces from the Lambda function, set the environment variable `DD_TRACE_ENABLED` to `false` on your own Lambda function.
1. If you are not interested in collecting logs from the Lambda function, update the Forwarder's CloudFormation stack parameter `DdForwardLog` to `false`.
1. Import and use the helper function from the Datadog Lambda Library, such as `lambda_metric` or `sendDistributionMetric`, to submit your custom metrics following the [sample code](#custom-metrics-sample-code).

> You can also emit metrics to CloudWatch logs in the following json format on your own from any Lambda runtime without installing the Datadog Lambda Library.

Reviewer 2: ```suggestion
You can also emit metrics to CloudWatch logs in the following JSON format on your own from any Lambda runtime without installing the Datadog Lambda Library:
```

}
```

{{< /programming-lang >}}
{{< /programming-lang-wrapper >}}

## Third-party libraries

> There are a number of open source libraries that make it easy to submit custom metrics to Datadog. However, many have not been updated to use [Distribution metrics][2], which are optimized for Lambda. Distribution metrics allow for server-side aggregations independent of a host, and give you flexible aggregations and tagging.

Reviewer: I'm not sure this section makes sense to include unless we are recommending some third-party libraries and linking to which third party libraries make this easy (which I'm not sure makes sense to do?)

Author: Yeah, agree, will remove.
